src/java.base/share/classes/java/util/concurrent/CompletableFuture.java

```diff
@@ -2925,48 +2929,23 @@ public static <U> CompletableFuture<U> failedFuture(Throwable ex) {
      * @since 9
      */
     public static <U> CompletionStage<U> failedStage(Throwable ex) {
-        if (ex == null) throw new NullPointerException();
-        return new MinimalStage<U>(new AltResult(ex));
-    }
-
-    /**
-     * Singleton delay scheduler, used only for starting and
-     * cancelling tasks.
-     */
-    static final class Delayer {
-        static ScheduledFuture<?> delay(Runnable command, long delay,
-                                        TimeUnit unit) {
-            return delayer.schedule(command, delay, unit);
-        }
-
-        static final class DaemonThreadFactory implements ThreadFactory {
-            public Thread newThread(Runnable r) {
-                Thread t = new Thread(r);
-                t.setDaemon(true);
-                t.setName("CompletableFutureDelayScheduler");
-                return t;
-            }
-        }
-
-        static final ScheduledThreadPoolExecutor delayer;
-        static {
-            (delayer = new ScheduledThreadPoolExecutor(
-                1, new DaemonThreadFactory())).
-                setRemoveOnCancelPolicy(true);
-        }
+        return new MinimalStage<U>(new AltResult(Objects.requireNonNull(ex)));
     }
 
     // Little class-ified lambdas to better support monitoring
 
     static final class DelayedExecutor implements Executor {
-        final long delay;
-        final TimeUnit unit;
+        final long nanoDelay;
         final Executor executor;
-        DelayedExecutor(long delay, TimeUnit unit, Executor executor) {
-            this.delay = delay; this.unit = unit; this.executor = executor;
+        DelayedExecutor(long nanoDelay, Executor executor) {
+            this.nanoDelay = nanoDelay; this.executor = executor;
         }
         public void execute(Runnable r) {
-            Delayer.delay(new TaskSubmitter(executor, r), delay, unit);
+            ForkJoinPool e = ASYNC_POOL;  // Use immediate mode to relay task
+            e.scheduleDelayedTask(
+                new ScheduledForkJoinTask<Void>(
+                    nanoDelay, 0L, true,
```

Comment:
@DougLea Doesn't this mean that anyone can submit tasks which will be executed on the DelayedScheduler scheduling thread? If so, there's a big risk that things end up there which shouldn't be there, and could even compromise the liveness of said thread? 🤔

Reply:
The task here is just TaskSubmitter's { executor.execute(action); }, which is assumed here (and in every other async  CF method) not to indefinitely block or loop. The class-level javadoc doesn't (and can't) strictly mandate this though.